src/software/util/design_patterns/constructor_arg_generic_factory.h

```diff
@@ -0,0 +1,187 @@
+     *
+     * @return a list of creator functions that are registered in this factory
+     */
+    static std::vector<std::function<std::unique_ptr<TypeToCreate>(std::any)>>
+    getRegisteredConstructors();
+
+   protected:
+    /**
+     * Adds a generic creator function to the registry
+     *
+     * @param generic_name The name of the generic creator function to be added
+     * @param generic_creator A "create" function that takes no arguments and will return
+     * a unique_ptr to a new instance of the specified generic type
+     */
+    static void registerCreator(
+        std::string generic_name,
+        std::function<std::unique_ptr<TypeToCreate>(std::any)> generic_creator);
+
+   private:
+    /**
+     * Returns a reference to the generic registry. The registry is a map of Generic names
+     * to a "create" function that will create and return a unique_ptr to a new concrete
+     * instance of the generic type, which allows the code to be aware
+     * of all the Generics that are available.
+     *
+     * This is the same as the above public getRegistry function. We need a mutable
+     * version in order to add entries to the registry. The function is private so that
+     * only this class can make the modifications. Outside sources should not have direct
+     * access to modify the registry.
+     *
+     * @return a mutable reference to the generic registry
+     */
+    static ConstructorArgGenericRegistry<IndexType, TypeToCreate>& getMutableRegistry();
+};
+
+
+template <class IndexType, class TypeToCreate, class T, class ArgType>
+class TConstructorArgGenericFactory : public ConstructorArgGenericFactory<IndexType, TypeToCreate>
+{
+    // compile time type checking that T is derived class of Generic
+    static_assert(std::is_base_of<TypeToCreate, T>::value,
+                  "T must be derived class of TypeToCreate!");
+
+   public:
+    TConstructorArgGenericFactory()
+    {
+        auto generic_creator = [](std::any arg) -> std::unique_ptr<TypeToCreate> {
+            std::optional<ArgType> constructor_arg_or_null;
+
+            try {
+                constructor_arg_or_null = std::any_cast<ArgType>(arg);
+            } catch (const std::bad_any_cast& error) {
+                std::stringstream error_ss;
+                error_ss << "Argument type is wrong! Could not any_cast "
+                         << arg.type().name() << " to ArgType " << typeid(T).name();
+                throw std::invalid_argument(error_ss.str());
+            }
+
+            return std::make_unique<T>(*constructor_arg_or_null);
+        };
+        ConstructorArgGenericFactory<IndexType, TypeToCreate>::registerCreator(TYPENAME(T),
+                                                                 generic_creator);
+    }
+};
+
+
+template <class IndexType, typename TypeToCreate, typename ConstructorType>
+class TMultipleConstructorArgGenericFactory : public ConstructorArgGenericFactory<IndexType, TypeToCreate>
```

Comment:
Definitely would be good to have a jdoc with a usage example here.

Reply:
added to the beginning of this file